### PR TITLE
[jit] Use computed goto to dispatch to next instruction in interpreter.

### PR DESCRIPTION
**Pull request by zhxchen17**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60211 [jit] Use computed goto to dispatch to next instruction in interpreter.**

Summary:
Refactor the interpreter loop a bit to use `goto` to jump to the next
instruction at the end of each instruction body, only on supported compiler.
In some microbenchmark this can reduce the branch misprediction rate of
interpreter loop by 10x. Altough there isn't much improvement for running
time in real cases, it's potentially still beneficial to do this optimization
and refactor the loop to be more readable in the meanwhile.

Test Plan:
CI

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29211283](https://our.internmc.facebook.com/intern/diff/D29211283)

### CODE REVIEW

**Comment by zhxchen17** [comment]
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29211283).

**Comment by zhxchen17** [comment]
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29211283).

**Comment by zhxchen17** [comment]
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29211283).

File: torch/csrc/jit/runtime/interpreter.cpp [hunk 1]
```diff
@@ -208,6 +208,37 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
     checkAndStartRecordFunction(frames.back(), stack);
   }
 
+#if defined(__GNUC__) || defined(__clang__)
+#define USE_COMPUTED_GOTO
```
Comment (gmagogsfm) [inline torch/csrc/jit/runtime/interpreter.cpp:212]: Nit:
Should we use a more specific name like JIT_USE_COMPUTED_GOTO and `undef` it at end of this logic block?

File: torch/csrc/jit/runtime/interpreter.cpp [hunk 2]
```diff
@@ -208,6 +208,37 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
...
+// We maintain two local variables as the internal interpreter state:
+// `frame` will be the current frame that the interpreter operatos on.
+// `inst` will the current instruction pointed to by program counter.
+//
+// Instruction blocks should be always declared through `INST` macro and
+// the instruction body should always start with a `INST_GUARD` declaration.
+// Also blocks should be ended properly with either `INST_NEXT` (for going
+// to the next instruction), or `INST_DISPATCH` (for jumping to a computed
+// position using `INST_FETCH`).
+#define INST_FETCH(X) (frame.function->instructions_[frame.pc += (X)])
```
Comment (gmagogsfm) [inline torch/csrc/jit/runtime/interpreter.cpp:224]: Optional:
Maybe code will be cleaner if change `INST_FETCH(X)` to:

`inst = (frame.function->instructions_[frame.pc += (X)])`

**Review by gmagogsfm — APPROVED** [review]
Looks good, only some optional comments

**Comment by zhxchen17** [comment]
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29211283).

**Pull request closed without merging**